Makefile

```diff
@@ -1,45 +1,96 @@
+TEST_PLAYBOOK ?= test.yml
+VAGRANT_BOX ?= ubuntu/trusty64
+PATH := $(PWD)/.venv_$(ANSIBLE_INSTALL_VERSION)/bin:$(shell printenv PATH)
+SHELL := env PATH=$(PATH) /bin/bash
 
 .DEFAULT_GOAL := help
 .PHONY: help
 
-all: test vagrant_halt clean
+export ANSIBLE_CONFIG
+export PATH
+export VAGRANT_BOX
+export TEST_PLAYBOOK
+
+all: test clean
 
 ## Run tests on any file change
 watch: test_deps
 	while sleep 1; do \
-		find defaults/ meta/ tasks/ templates/ tests/test.yml tests/vagrant/Vagrantfile \
+		find defaults/ handlers/ meta/ tasks/ templates/ tests/test.yml \
 		| entr -d make lint vagrant; \
 	done
 
 ## Run tests
-test: test_deps lint vagrant
+test: lint test_deps vagrant
 
 ## Install test dependencies
-test_deps:
-	rm -rf tests/sansible.*
-	ln -s .. tests/sansible.gocd_agent
-	ansible-galaxy install --force -p tests/ -r tests/local_requirements.yml
+test_deps: .venv_$(ANSIBLE_INSTALL_VERSION) tests/roles
+
+tests/roles:
+	mkdir -p tests/roles
+	ln -s ../.. tests/roles/sansible.$(ROLE_NAME)
+	ansible-galaxy install -p tests/roles -r tests/local_requirements.yml --ignore-errors
+
+## ! Executes Ansible tests using local connection
+# run it ONLY from within a test VM.
+# If you want to test this role, run `make test` instead.
+# Example: make test_ansible
+#          make test_ansible TEST_PLAYBOOK=test-something-else.yml
+test_ansible: test_ansible_build test_ansible_configure
+	cd tests && ansible-playbook \
+		--inventory inventory \
+		--connection local \
+		--tags assert \
+		$(TEST_PLAYBOOK)
+
+## ! Executes Ansible tests using local connection
+# run it ONLY from witinh a test VM.
+# Example: make test_ansible_build
+#          make test_ansible_build TEST_PLAYBOOK=test-something-else.yml
+test_ansible_%:
+	cd tests && ansible-playbook \
+		--inventory inventory \
+		--connection local \
+		--tags=$(subst test_ansible_,,$@) \
+		$(TEST_PLAYBOOK)
+	cd tests && ansible-playbook \
+		--inventory inventory \
+		--connection local \
+		--tags=$(subst test_ansible_,,$@) \
+		$(TEST_PLAYBOOK) \
+		| grep -q 'changed=0.*failed=0' \
+			&& (echo 'Idempotence test: pass' && exit 0) \
+			|| (echo 'Idempotence test: fail' && exit 1)
 
 ## Start and (re)provisiom Vagrant test box
 vagrant:
-	cd tests/vagrant && vagrant up --no-provision
-	cd tests/vagrant && vagrant provision
+	cd tests && vagrant up --no-provision
+	cd tests && vagrant provision
+	@echo "- - - - - - - - - - - - - - - - - - - - - - -"
+	@echo "           Provisioning Successful"
+	@echo "- - - - - - - - - - - - - - - - - - - - - - -"
 
 ## Execute simple Vagrant command
 # Example: make vagrant_ssh
 #          make vagrant_halt
 vagrant_%:
-	cd tests/vagrant && vagrant $(subst vagrant_,,$@)
+	cd tests && vagrant $(subst vagrant_,,$@)
 
-## Lint role
-# You need to install ansible-lint
-lint:
-	find defaults/ meta/ tasks/ templates/ -name "*.yml" | xargs -I{} ansible-lint {}
+## Installs a virtual environment and all python dependencies
+.venv_%:
```

Comment:
ansible lint and ansible version.

You can now test different version locally if needed.

`make test ANSIBLE_INSTALL_VERSION=2.3.1.0`